### PR TITLE
WIP: Localization of the ffmpeg compatibility macros

### PR DESCRIPTION
**Pull request by VadimLevin**
The general idea of this patch is to localize deprecated functionality from the FFMPEG libraries into the `compat` functions and make code cleaner and easier to maintain.

Some of the parts and workflows are updated according to the FFMPEG recommended functionality e.g. `avcodec_close` is changed to `avcodec_free_context`.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

<cut/>

```
force_builders=Custom
buildworker:Custom=linux-1,linux-2,linux-4
build_image:Custom=ffmpeg-master
```

### CODE REVIEW

File: modules/videoio/src/cap_ffmpeg_impl.hpp [hunk 1]
```diff
@@ -2281,186 +2423,71 @@ bool CvVideoWriter_FFMPEG::open( const char * filename, int fourcc,
...
+        };
+        if ((codec_id = av_codec_get_id(fallback_tags, fourcc)) == CV_CODEC(CODEC_ID_NONE))
         {
             fflush(stdout);
-            fprintf(stderr, "OpenCV: FFMPEG: tag 0x%08x/'%c%c%c%c' is not found (format '%s / %s')'\n",
-                    fourcc, CV_TAG_TO_PRINTABLE_CHAR4(fourcc),
-                    fmt->name, fmt->long_name);
+            fprintf(stderr,
+                    "OpenCV: FFMPEG: tag 0x%08x/'%c%c%c%c' is not found (format '%s / %s')'\n",
+                    fourcc, CV_TAG_TO_PRINTABLE_CHAR4(fourcc), fmt->name, fmt->long_name);
```
Comment (alalek) [inline modules/videoio/src/cap_ffmpeg_impl.hpp]: Please avoid playing with whitespaces.

Patch is huge (>100 lines), so don't make it worse for review.

File: modules/videoio/src/cap_ffmpeg_impl.hpp [hunk 2]
```diff
@@ -3195,7 +3104,7 @@ bool InputMediaStream_FFMPEG::read(unsigned char** data, int* size, int* endOfFi
 
         if (pkt_.stream_index != video_stream_id_)
         {
-            _opencv_ffmpeg_av_packet_unref(&pkt_);
+            compat::avPacketUnref(&pkt_);
```
Comment (alalek) [inline modules/videoio/src/cap_ffmpeg_impl.hpp]: Please keep original function identifier in names (of the latest modern FFmpeg version).:
- original: av_packet_unref
- good compatibility wrapper name: `_opencv_ffmpeg_av_packet_unref`
- avoid that: `compat::avPacketUnref`

Approach:
```
... define _opencv_ffmpeg_av_packet_unref ...

#define av_packet_unref   av_packet_unref_must_be_wrapped  // guard

... other code without av_packet_unref() ...
```

File: modules/videoio/src/cap_ffmpeg_impl.hpp [hunk 3]
```diff
@@ -1143,45 +1297,20 @@ bool CvCapture_FFMPEG::processRawPacket()
...
-#if LIBAVFORMAT_BUILD > 4628
-        AVCodecContext* ctx = ic->streams[video_stream]->codec;
-#else
-        AVCodecContext* ctx = &ic->streams[video_stream]->codec;
-#endif
-        int err = av_bitstream_filter_filter(bsfc, ctx, NULL, &packet_filtered.data,
-            &packet_filtered.size, packet.data, packet.size, packet_filtered.flags & AV_PKT_FLAG_KEY);
-        if (err < 0)
+        if (!compat::avBitstreamFilterFilter(bsfc, ic->streams[video_stream]->codec, packet,
+                                             packet_filtered))
```
Comment (alalek) [inline modules/videoio/src/cap_ffmpeg_impl.hpp]: Avoid extraction of one time used code into functions.

File: modules/videoio/src/cap_ffmpeg_impl.hpp (hunk 2 above)
Comment (VadimLevin) [reply modules/videoio/src/cap_ffmpeg_impl.hpp]: I've tried to place all functions to `compatibility` namespace with pascal case naming (keeping the original naming).
But I've got your point and change it to what you suggest. Thank you.

File: modules/videoio/src/cap_ffmpeg_impl.hpp (hunk 1 above)
Comment (VadimLevin) [reply modules/videoio/src/cap_ffmpeg_impl.hpp]: I have enabled autoformat, sorry.

File: modules/videoio/src/cap_ffmpeg_impl.hpp (hunk 3 above)
Comment (VadimLevin) [reply modules/videoio/src/cap_ffmpeg_impl.hpp]: I tried to remove versions `#ifdef`'s from the actual functions of the capture and writer, to keep them simple and clean. What is the matter to do not do it?

**Comment by VadimLevin** [comment]
> Main question here is what changes are really necessary?

The general points are: 
* Make functions simple and clean with removing `#ifdef`'s from the body of the functions, those do actual work. 
* Remove old functionality ([FFmpeg libs with versions < 0.5](https://github.com/FFmpeg/FFmpeg/tree/release/0.5))
* Fix wrong encoder/decoder initializations/freeing (e.g. do not use [`avcodec_get_context_defaults3`](https://github.com/FFmpeg/FFmpeg/commit/9b211c43dc5f2e618f204c4a7fd184eb2ea51f02))
* Add missing memory deallocations

Should I split this into several PR, can you suggest what is more important?

**Pull request closed without merging**